### PR TITLE
UI Textbox

### PR DESCRIPTION
**Pull request by jytang**
- made things that had text + a background color into a UITextbox.

### CODE REVIEW

**Review by AznCelebration — APPROVED** [review]

File: src/ui/ui_text_box.cpp [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+                     float alpha) :
+    UIContainer(start_x, start_y),
+    width(width), height(height),
+    alpha(alpha), initial_alpha(alpha) {
+
+    bg = UIRectangle(0, 0, width, height, bg_color, alpha);
+    attach(bg);
+
+    // Set text node accordingly, fill bg as much as possible.
+    float text_width = width / text.length() / X_SCALE_FACTOR;
```
Comment (metakirby5) [inline src/ui/ui_text_box.cpp]: Parens would be nice.

File: src/ui/ui_text_box.cpp [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    float text_height = height / Y_SCALE_FACTOR;
+    float y_offset = height / Y_OFFSET_FACTOR;
+    text_node = UITextNode(text, 0, y_offset, text_width, text_height, text_color);
+    attach(text_node);
+
+    // Setup debug listener
+    events::debug::toggle_ui_text_box_background_event.connect([&]{
+        this->alpha = this->alpha == 0.99f ? initial_alpha : 0.99f;
```
Comment (metakirby5) [inline src/ui/ui_text_box.cpp]: I heard float equality comparisons are bad. Maybe this is okay though...

File: src/ui/ui_text_box.h [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+#include "ui_container.h"
+#include "ui_text_node.h"
+#include "ui_rectangle.h"
+#include "util/color.h"
+#include <string>
+
+class UITextBox : public UIContainer {
+public:
+    UITextBox() {} // default ctor
```
Comment (metakirby5) [inline src/ui/ui_text_box.h]: Is it wise to provide this? May be better to force usage of the parameterized constructor.

File: src/ui/ui_text_box.h [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    UITextBox(std::string text,
+              float start_x, float start_y,
+              float width, float height,
+              Color text_color, Color bg_color = Color::WHITE,
+              float alpha = 0.f);
+    void set_text(std::string new_text);
+private:
+    float width, height;
```
Comment (metakirby5) [inline src/ui/ui_text_box.h]: Consider adding setters/getters for these which also update the text scale, etc. accordingly.

**Review by metakirby5 — CHANGES_REQUESTED** [review]
nic

**Pull request merged**